src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+                if (context.UseSspi)
+                { 
+                    if (isAsync)
+                    {
+                        await stream.WriteAsync(outSSPIBuff.AsMemory(0, (int)outSSPILength), ct).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        stream.Write(outSSPIBuff.AsSpan(0, (int)outSSPILength));
+                    }
+                }
+                await stream.WriteStringAsync(context.AttachedDbFileName, isAsync, ct).ConfigureAwait(false);
+                if (!context.UseSspi && !(context.Features.FederatedAuthenticationInfoRequested || context.Features.FederatedAuthenticationRequested))
+                {
+                    if (context.PasswordChangeRequest?.NewSecurePassword != null)
+                    {
+                        // TODO : implement saving secure string.
+                        throw new NotImplementedException();
+                        // _physicalStateObj.WriteSecureString(rec.newSecurePassword);
+                    }
+                    else
+                    {
+                        if (isAsync)
+                        {
+                            await stream.WriteAsync(encryptedChangePassword.AsMemory(0, encryptedChangePasswordLengthInBytes), ct).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            stream.Write(encryptedChangePassword.AsSpan(0, encryptedChangePasswordLengthInBytes));
+                        }
+                    }
+                }
+
+                await SendFeatureExtensionData(context, 
+                    requestedFeatures, 
+                    recoverySessionData, 
+                    fedAuthFeatureExtensionData, 
+                    isAsync, 
+                    ct).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                if (ADP.IsCatchableExceptionType(e))
+                {
+                    // Reset the buffer if there was an exception.
+                    context.TdsStream.Reset();
+                }
+
+                throw;
+            }
+        }
+
+        private async ValueTask SendFeatureExtensionData(LoginHandlerContext context,
+                                        TdsEnums.FeatureExtension requestedFeatures,
+                                        SessionData recoverySessionData,
+                                        FederatedAuthenticationFeatureExtensionData fedAuthFeatureExtensionData,
+                                        bool isAsync,
+                                        CancellationToken ct)
+        {
+            if (context.UseFeatureExt)
+            {
+                if (_sessionRecoveryFeature.ShouldUseFeature(requestedFeatures))
+                {
+                    await _sessionRecoveryFeature.WriteFeatureData(context, isAsync, ct).ConfigureAwait(false);
```

Comment:
Since you're already passing context, can the WriteFeatureData not determine itself whether this feature was requested? I wonder if ShouldUseFeature could be possibly removed and made internal to WriteFeatureData?

Reply:
In my mind these feature classes contain the logic for the work done on the connection, rather than contain information like "Should Use feature" which is contained with the connection. This follows the handler paradigm where the state is maintained in the context, and the word is done in these feature classes.
The state is essentially flowing with the context, which the handler decides on, to invoke the right actions using the feature classes. 

Based on this feedback, I am going to move the features to the context, and then these features can call "Should Use" and then when your Parser is ready, it can add information to these feature classes to add the relevant information to be saved with these features.